rasa/core/featurizers/precomputation.py

```diff
@@ -0,0 +1,414 @@
+        """Adds the given message if it is not already present.
+
+        Args:
+          message_with_one_key_attribute: The message we want to add to the lookup
+            table. It must have exactly one key attribute.
+
+        Raises:
+          `ValueError` if the given message does not contain exactly one key
+          attribute or if there is a collision with a message that has a different
+          hash value
+        """
+        # extract the key pair
+        attributes = message_with_one_key_attribute.data.keys()
+        key_attributes = set(attributes).intersection(self.KEY_ATTRIBUTES)
+        if not key_attributes or len(key_attributes) != 1:
+            raise ValueError(
+                f"Expected exactly one attribute out of "
+                f"{self.KEY_ATTRIBUTES} but received {len(attributes)} attributes "
+                f"({attributes})."
+            )
+        key_attribute = list(key_attributes)[0]
+        key_value = str(message_with_one_key_attribute.data[key_attribute])
+        # extract the message
+        existing_message = self._table[key_attribute].get(key_value)
+        if existing_message is not None:
+            if hash(existing_message) != hash(message_with_one_key_attribute):
+                raise ValueError(
+                    f"Expected added message to be consistent. "
+                    f"({key_attribute}, {key_value}) already maps "
+                    f"to {existing_message}, but we want to add "
+                    f"{message_with_one_key_attribute} now."
+                )
+            else:
+                self._num_collisions_ignored += 1
+        else:
+            self._table[key_attribute][key_value] = message_with_one_key_attribute
+
+    def add_all(self, messages_with_one_key_attribute: List[Message]) -> None:
+        """Adds the given messages.
+
+        Args:
+          messages_with_one_key_attribute: The messages that we want to add.
+            Each one must have exactly one key attribute.
+
+        Raises:
+          `ValueError` if we cannot create a key for the given message or if there is
+          a collisions with a message that has a different hash value
+        """
+        for message in messages_with_one_key_attribute:
+            self.add(message)
+
+    def collect_features(
+        self, sub_state: SubState, attributes: Optional[Iterable[Text]] = None
+    ) -> Dict[Text, List[Features]]:
+        """Collects features for all attributes in the given substate.
+
+        There might be be multiple messages in the container that contain features
+        relevant for the given substate, e.g. this is the case if `TEXT` and
+        `INTENT` are present in the given message. All of those messages will be
+        collected and their features combined.
+
+        Args:
+          sub_state: substate for which we want to extract the relevent features
+          attributes: if not `None`, this specifies the list of the attributes of the
```

Comment:
This can be arbitrary attributes. If the NLU pipeline has added attributes and features for these attributes, then it is possible that we want to list attributes here which are not key attributes. (Should not happen at the moment - we add new attributes (TOKENS) but the attribute field in the features should be "TEXT" 🤔  ... Good that you asked! I find this a bit confusing ... Guess we'll definitely add that to the documentation on the featurizers that we've just started   [here](https://www.notion.so/rasa/Featurizers-3de084fb437f4127814fe7f8f0a20fcc) )

Reply:
oh, no actually it is not that confusing because tokens are always stored as TOKEN_NAMES[<attribute>] and the features then just remember <attribute> :)  .... well, at least it is not confusing as long as we only have a single tokenizer :D